### ISSUE TITLE
Competition linked to stream

### ISSUE THREAD

**Issue by scd** [theCrag/website#3138]
**What happened?**

If ascents in a stream are associated with a competition leaderboard then the competition(s) should be shown in the stream page.

 - [x] SQL to detect association with a competition leaderboard - list of ascent ids returns competition ids. (@scd)
 - [x] Add competition summary details to stream event data parameters ( @scd )
 - [x] Make the UI really cool for linking to competitions in stream event (@brendanheywood )
 - [ ] /competition/12345/event/1745768796 end point
 - [x] fix up units (eg points)
 - [ ] use link_to - add options to return components of link vs html

Here is an example url in brendan.thecrag.com

https://brendan.thecrag.com/event/1017089796

**Comment by brendanheywood** [comment]
I think we need to be nuanced here, if we have a comp which is say easter for aus, then almost everyone who ticks stuff and who sees ticks in aus will be looking almost exclusively as aus ticks. So we don't want something in your face which is then repeated for every single stream event.

![image](https://user-images.githubusercontent.com/187449/38399378-6d794d52-398d-11e8-8a66-f4f08f7afb4d.png)

So I think this should replace the cpr / karma score for a stream event, and be really minimal:

![image](https://user-images.githubusercontent.com/187449/38399601-a89a3454-398e-11e8-9bf0-f76b6165b125.png)

I'm wondering if we can even get away with something a little more custom for each comp:

![image](https://user-images.githubusercontent.com/187449/38399802-e6072364-398f-11e8-82f5-c52ab3b3cb31.png)

**Comment by scd** [comment]
This is a nice observation. There are two challenges we need to get around and perhaps it does not really matter.

1. One ascent can belong to many comps, or even multiple leaderboards in a single comp. One stream event could be associated with multiple comps.

2. We cannot just sum the points across all the ascents. There is an algorithm assigned to the leaderboard so we can hook into that. However this algorithm can be quite complex (eg hangman) so there is potentially a performance issue if we have to collect and supply all the information for every stream event.

3. The use of the word `points` applies to a lot of leaderboards, but some of them have other units such as `meters`. 

If there was a design with the points on the actual ascent then a lot of issues would be simpler. In some ways it would be good to know which ascents qualified.

**Comment by brendanheywood** [comment]
I assumed that the points, and units would be supplied by the template data I wouldn't hard code them. Whatever logic you see in the comp ascents facet page should be identical to here. I guess one weird thing is that the points associated with an ascent could change over time, like if we only count the top 6 ascents. I think we should simply avoid every doing things like that.

Re multiple comps, I think we should generally avoid having more than one comp in the same location at the same time. If there is an overlap the stream would just show the one with the largest points but even that is hairy because point of 2 in one comp might actually be more useful in terms of winning then 10 points in another comp.

One stream having lots of ascents with each linked to only one comp, again should be a small edge case and I'd just pick the biggest comp, or even just not worry and pick the first one.

There is also the question of where you go when you click on the little badge, because this is designed for people who may be unaware of the comp I think this needs to take you to somewhere with lots of context, which explains not only what the comp is, but specifically the detailed points related to this stream event. I think just punting them to the comp page will be potentially confusing.

So for instance for this stream event:

https://www.thecrag.com/event/1745768796

We would link to a page maybe something like:

/competition/12345/event/1745768796

Or perhaps something specific to the leaderboard and climber and date:

/competition/12345/detail?climber=1234&board=2345&date=2018-04-06

Which would have a chunk of text and a few bits pulled from the comp info, and a directly embedded list of routes (not facet UI) with the points with the extra filter for the date like this:

https://www.thecrag.com/climbing/australia/ascents/has/cleanbst/between/2018-04-05+2018-04-05/with-route-gear-style/boulder/in-setting/natural/by/nc07287/?sortby=when,asc&leaderboard-score=1712391429

**Comment by brendanheywood** [comment]
Actually I think I would prefer:

/competition/12345/event/1745768796

And this page would potentially have multiple sets of ticks if there are multiple leaderboards. It also means the link itself can be a bit dumber.

**Comment by scd** [comment]
So this is a new end point. This will solve the problem of what to link to. 

We should not make any restrictions on multiple comps for a particular area. We could add a prioritisation field to comps and pick the highest priority and if equal pick the lowest id. This will solve the problem of which to show. 

I still want to be able to see my session CPR, which is clobbered by this design.

Note that the points should not change over time. They are a partial score for your overall account. Like CPR you cannot assume that that summing ascent scores or partial scores gives you an account score. You might get 5 points for the ascents in the event but this may add zero points to your overall account score (eg repeat ascent or hangman where you already have the letters).

**Comment by scd** [comment]
Can we toggle between badges?

**Comment by brendanheywood** [comment]
Like animate? Yeah but sounds ugly to me, it may also reflow the text around it which would be pretty bad. 

Lets just get the data ready and then we can play with some options showing both together. We'll also create some edge cases with multiple comps that overlap

**Comment by scd** [comment]
Yup agree. When we play with the solution I would like to be able to see the session CPR and one comp. I don't think we need to show all comps.

**Comment by scd** [comment]
I have done event data template for competition linkages:

```
"leaderboards": [
{
"leaderboard": {
"id": "618133861"
},
"competition": {
"name": "Easter Tick Contest",
"stub": "2018-aussie-easter-tick-contest",
"id": "618910366",
"type": "competition"
},
"score": 4
}
],
```

**Comment by scd** [comment]
Example url

https://brendan.thecrag.com/event/1017089796

**Comment by brendanheywood** [comment]
Ok this is mostly working but just needs a few loose ends with the template data sorted:

1) I'm given a leaderboard id, ie 

```
"leaderboard": {
   "id": "1014701964"
}
```

but in the inflate data there is no id present to match against:

```
"leaderboards": [
 {
    "url": "/climbing/south-korea/ascents/between/2016-11-12+2017-04-12/sortby=when,desc",
    "title": "Everybody",
    "itemScore": "count",
    "unit": "point",
    "unitHeader": "Points",
    "activity": "ascent"
  }
],
```
It needs the id as an attribute. The only reason I need this is for the units 'Points' so this could just be added to the event rather than the inflate.

2) I couldn't get this working with the link_to, I think for this case and many cases we need a more raw function which just returns the url, or possibly an object with the url, label, title etc so we can pull it back out.

**Comment by scd** [comment]
I have updated units to come from templated data, translated and pluralised.

**Comment by brendanheywood** [comment]
At the moment the badge just links to the comp page and we could just leave it at that?

**Comment by scd** [comment]
Yes I am happy leaving it like this for a while. The main driver for me is to show the comp in the streams. Other linkages are something we need to do, but can be done later.

**Comment by brendanheywood** [comment]
Actually I think the language prefix url is a must have for may release given the translated comp?
